### PR TITLE
auth-backend-module-microsoft-provider: update scope handling to define required scopes in backend

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Following #24743 the Microsoft provider was left in a bit of an awkward state, with the `additionalScopes` config being respected but included even when requesting scopes for a resource. This refactors the Microsoft provider to use the new `scopes` config and declaring the required scopes in the backend, but also transforming the request scopes to account for the resource scope separation that's managed in the frontend here: https://github.com/backstage/backstage/blob/b645d70034a05b424d63f3e7ef8a64638d3733a3/packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts#L93-L156

Think it could be good to have some eyes on this from some Microsoft users: @afscrome @TheGemmell @pjungermann @awanlin 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by awanlin — APPROVED** [review]

File: plugins/auth-backend-module-microsoft-provider/src/authenticator.ts [hunk 1]
```diff
@@ -21,21 +21,30 @@ import {
   PassportProfile,
 } from '@backstage/plugin-auth-node';
 import { ExtendedMicrosoftStrategy } from './strategy';
-import { union } from 'lodash';
```
Comment (benjdlambert) [inline plugins/auth-backend-module-microsoft-provider/src/authenticator.ts:24]: Can also :wave: bye bye to lodash

**Review by benjdlambert — APPROVED** [review]

**Comment by jpitapeva** [comment]
Hi, i did the install today for 1.30 version by (npx @backstage/create-app@latest) and show the same error, can you help me?

Error
`Login failed; caused by Error: The microsoft provider is not configured to support sign-in`

---

file: packages/app/src/App.tsx

import { microsoftAuthApiRef } from '@backstage/core-plugin-api';
```
components: {
    SignInPage: props => <SignInPage {...props} 
    auto providers={[
      {
        id: 'microsoft-auth-provider',
        title: 'microsoft',
        message: 'Sign in using microsoft',
        apiRef: microsoftAuthApiRef,
      },
    
    ]} 
    
```

---

file: packages/backend/src/index.ts

backend.add(import('@backstage/plugin-auth-backend-module-microsoft-provider'));
```

---

file: app-config.yaml

auth:
  environment: development
  providers:
    microsoft:
      development:
        clientId: "REMOVED"
        clientSecret: "REMOVED"
        tenantId: "REMOVED"
        domainHint: "REMOVED"
        signIn:
           resolvers:
             # typically you would pick one of these
             - resolver: emailMatchingUserEntityProfileEmail
             - resolver: emailLocalPartMatchingUserEntityName
             - resolver: emailMatchingUserEntityAnnotation

File: plugins/auth-backend-module-microsoft-provider/src/authenticator.ts [hunk 2]
```diff
@@ -21,21 +21,30 @@ import {
...
   defaultProfileTransform:
     PassportOAuthAuthenticatorHelper.defaultProfileTransform,
+  scopes: {
+    required: ['email', 'openid', 'offline_access', 'user.read'],
+    transform({ requested, granted, required, additional }) {
+      // Resources scopes are of the form `<resource>/<scope>`, and are handled
+      // separately from the normal scopes in the client. When request a
+      // resource scope we should only include forward the request scope along
+      // with offline_access.
+      const hasResourceScope = Array.from(requested).some(s => s.includes('/'));
```
Comment (TheGemmell) [inline plugins/auth-backend-module-microsoft-provider/src/authenticator.ts:36]: Did some testing locally.
The additionalScopes config value gets passed into 'additional' not 'requested'

This is tested as working when directly modifying the node_module locally, small change:
```
const hasResourceScope = Array.from(additional).some((s) => s.includes("/"));
if (hasResourceScope) {
      return [...additional, "offline_access"];
}
```

Comment (TheGemmell) [reply plugins/auth-backend-module-microsoft-provider/src/authenticator.ts:36]: @Rugvip 
Should I make an issue about this?

**Pull request merged**